File: everything_parsing/06-ambiguity-disambiguation/java/AmbiguityTest.java
import java.util.List;

/** java -cp java/out AmbiguityTest */
public final class AmbiguityTest {
    public static void main(String[] args) {
        System.out.println("AmbiguityTest");

        Grammar amb = Grammar.parse("E -> E + E | E * E | ( E ) | id");
        Grammar unamb = Grammar.parse("""
            E -> E + T | T
            T -> T * F | F
            F -> ( E ) | id
            """);

        List<String> mul = List.of("id", "+", "id", "*", "id");
        List<String> add = List.of("id", "+", "id", "+", "id");
        List<String> one = List.of("id");
        List<String> two = List.of("id", "+", "id");

        // ── the ambiguous grammar ──
        Assert.equals(Ambiguity.allTrees(amb, one).size(), 1, "id has one tree even in the ambiguous grammar");
        Assert.equals(Ambiguity.allTrees(amb, two).size(), 1, "id + id has one tree (only one operator)");
        Assert.equals(Ambiguity.allTrees(amb, mul).size(), 2, "id + id * id has TWO trees");
        Assert.equals(Ambiguity.allTrees(amb, add).size(), 2, "id + id + id has TWO trees");
        Assert.that(Ambiguity.isAmbiguousFor(amb, mul), "grammar is ambiguous for id + id * id");

        // the two trees for id + id * id mean different numbers (id = 2)
        var t = Ambiguity.allTrees(amb, mul);
        long v0 = Main.evalExpr(t.get(0));
        long v1 = Main.evalExpr(t.get(1));
        Assert.that((v0 == 6 && v1 == 8) || (v0 == 8 && v1 == 6),
            "the two trees evaluate to 6 and 8");
        Assert.that(v0 != v1, "ambiguity here means two different values");

        // the two trees for id + id + id mean the SAME number but group differently
        var a = Ambiguity.allTrees(amb, add);
        Assert.equals(Main.evalExpr(a.get(0)), 6L, "left-assoc value");
        Assert.equals(Main.evalExpr(a.get(1)), 6L, "right-assoc value (addition is associative)");
        Assert.that(!Main.grouping(a.get(0)).equals(Main.grouping(a.get(1))),
            "but the groupings differ");

        // ── the disambiguated grammar ──
        Assert.equals(Ambiguity.allTrees(unamb, mul).size(), 1, "disambiguated: id + id * id has ONE tree");
        Assert.equals(Ambiguity.allTrees(unamb, add).size(), 1, "disambiguated: id + id + id has ONE tree");
        Assert.equals(Main.grouping(Ambiguity.allTrees(unamb, mul).get(0)), "(id + (id * id))",
            "disambiguated grammar groups * before +");
        Assert.equals(Main.grouping(Ambiguity.allTrees(unamb, add).get(0)), "((id + id) + id)",
            "disambiguated grammar makes + left-associative");

        // ── same language ──
        Assert.equals(Derivation.enumerate(amb, 7), Derivation.enumerate(unamb, 7),
            "ambiguous and disambiguated grammars generate the same language (up to length 7)");

        // ── smallest ambiguous string ──
        Assert.equals(Ambiguity.smallestAmbiguousString(amb, 7), List.of("id", "*", "id", "*", "id"),
            "smallest ambiguous string is id * id * id");
        Assert.that(Ambiguity.smallestAmbiguousString(unamb, 9) == null,
            "the disambiguated grammar has no ambiguous string up to length 9");

        // ── dangling else ──
        Grammar dang = Grammar.parse("S -> if x then S | if x then S else S | a | b");
        var ie = List.of("if", "x", "then", "if", "x", "then", "a", "else", "b");
        Assert.equals(Ambiguity.allTrees(dang, ie).size(), 2, "dangling else: two trees");

        Assert.summary();
    }
}

File: everything_parsing/06-ambiguity-disambiguation/js/ambiguity.test.mjs
// node js/ambiguity.test.mjs

import { that, equals, summary } from "./assert.mjs";
import { Grammar } from "./grammar.mjs";
import { enumerate } from "./derivation.mjs";
import { allTrees, isAmbiguousFor, smallestAmbiguousString } from "./ambiguity.mjs";

console.log("ambiguity.test.mjs");

const amb = Grammar.parse("E -> E + E | E * E | ( E ) | id");
const unamb = Grammar.parse("E -> E + T | T\nT -> T * F | F\nF -> ( E ) | id\n");

const mul = ["id", "+", "id", "*", "id"];
const add = ["id", "+", "id", "+", "id"];

// ── the ambiguous grammar ──
equals(allTrees(amb, ["id"]).length, 1, "id has one tree");
equals(allTrees(amb, ["id", "+", "id"]).length, 1, "id + id has one tree");
equals(allTrees(amb, mul).length, 2, "id + id * id has TWO trees");
equals(allTrees(amb, add).length, 2, "id + id + id has TWO trees");
that(isAmbiguousFor(amb, mul), "grammar is ambiguous for id + id * id");

const t = allTrees(amb, mul);
const v0 = evalExpr(t[0]), v1 = evalExpr(t[1]);
that((v0 === 6 && v1 === 8) || (v0 === 8 && v1 === 6), "the two trees evaluate to 6 and 8");
that(v0 !== v1, "ambiguity here means two different values");

const a = allTrees(amb, add);
equals(evalExpr(a[0]), 6, "left-assoc value");
equals(evalExpr(a[1]), 6, "right-assoc value (addition is associative)");
that(grouping(a[0]) !== grouping(a[1]), "but the groupings differ");

// ── the disambiguated grammar ──
equals(allTrees(unamb, mul).length, 1, "disambiguated: id + id * id has ONE tree");
equals(allTrees(unamb, add).length, 1, "disambiguated: id + id + id has ONE tree");
equals(grouping(allTrees(unamb, mul)[0]), "(id + (id * id))", "disambiguated groups * before +");
equals(grouping(allTrees(unamb, add)[0]), "((id + id) + id)", "disambiguated makes + left-associative");

// ── same language ──
equals(enumerate(amb, 7).join("|"), enumerate(unamb, 7).join("|"),
  "ambiguous and disambiguated grammars generate the same language (length <= 7)");

// ── smallest ambiguous string ──
equals(smallestAmbiguousString(amb, 7).join(" "), "id * id * id", "smallest ambiguous string");
that(smallestAmbiguousString(unamb, 9) === null, "disambiguated grammar: none up to length 9");

// ── dangling else ──
const dang = Grammar.parse("S -> if x then S | if x then S else S | a | b");
const ie = ["if", "x", "then", "if", "x", "then", "a", "else", "b"];
equals(allTrees(dang, ie).length, 2, "dangling else: two trees");

summary();

function evalExpr(n) {
  if (!n.isExpanded) return n.symbol === "id" ? 2 : 0;
  const k = n.children;
  if (k.length === 1) return evalExpr(k[0]);
  if (k.length === 3) {
    if (k[0].symbol === "(") return evalExpr(k[1]);
    const x = evalExpr(k[0]), y = evalExpr(k[2]);
    if (k[1].symbol === "+") return x + y;
    if (k[1].symbol === "*") return x * y;
  }
  return 0;
}
function grouping(n) {
  if (!n.isExpanded) return n.symbol;
  const k = n.children;
  if (k.length === 1) return grouping(k[0]);
  if (k.length === 3) {
    if (k[0].symbol === "(") return "(" + grouping(k[1]) + ")";
    return "(" + grouping(k[0]) + " " + k[1].symbol + " " + grouping(k[2]) + ")";
  }
  return "?";
}
